jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -1530,28 +1552,146 @@ Method getMethodHelper(
+				// bestCandidate and result have same declaringClass.
+				Class<?> candidateRetType = bestCandidate.getReturnType();
+				Class<?> resultRetType = result.getReturnType();
+				if ((candidateRetType != resultRetType) && candidateRetType.isAssignableFrom(resultRetType)) {
+					bestCandidate = result;
+				}
 			}
 		}
 	}
 	return cacheMethod(bestCandidate);
 }
 
+/**
+ * Helper method searches all interfaces implemented by superclasses from the top to the bottom
+ * for the most specific method declared in one of these interfaces.
+ *
+ * @param currentClass the class to be searched, including the current class and all superclasses
+ * @param name the specified method's name
+ * @param parameterTypes the types of the arguments of the specified method
+ * @return the most specific method selected from all interfaces from each superclass of the current class;
+ *         otherwise, return the method of the first interface from the top superclass
+ *         if the return types of all specified methods are identical.
+ */
+private Method getMostSpecificMethodFromAllInterfacesOfAllSuperclasses(Class<?> currentClass, HashMap<Class<?>, HashMap<MethodInfo, MethodInfo>> infoCache, 
+	String name, Class<?>... parameterTypes) 
+{
+	Method candidateMethod = null;
+	if (currentClass != Object.class) {
+		/* get to the top superclass first. if all return types end up being the same the interfaces from this superclass have priority. */
+		candidateMethod = getMostSpecificMethodFromAllInterfacesOfAllSuperclasses(currentClass.getSuperclass(), infoCache, name, parameterTypes);
+		
+		/* search all interfaces of current class, comparing against result from previous superclass. */
+		candidateMethod = getMostSpecificMethodFromAllInterfacesOfCurrentClass(currentClass, infoCache, candidateMethod, name, parameterTypes);
+	}
+	return candidateMethod;
+}
+
+/**
+ * Helper method searches all interfaces implemented by the current class or interface 
+ * for the most specific method declared in one of these interfaces.
+ *
+ * @param currentClass the class or interface to be searched
+ * @param potentialCandidate potential candidate from superclass, null if currentClass is an interface
+ * @param name the specified method's name
+ * @param parameterTypes the types of the arguments of the specified method
+ * @return the most specific method selected from all interfaces;
+ *         otherwise if return types from all qualifying methods are identical, return an arbitrary method.
+ */
+private Method getMostSpecificMethodFromAllInterfacesOfCurrentClass(Class<?> currentClass, HashMap<Class<?>, HashMap<MethodInfo, MethodInfo>> infoCache,
+	Method potentialCandidate, String name, Class<?>... parameterTypes) 
+{
+	Method bestMethod = potentialCandidate;
+	/* if infoCache is passed in, reuse from superclass */
+	if (null == infoCache) {
+		infoCache = new HashMap<>(16);
+	}
+	HashMap<MethodInfo, MethodInfo> methodCandidates = getMethodSet(infoCache, false, true);
```

Comment:
is there a chance that `infoCache` could be confused by `localInterfacesOnly` since the key is always `this` class? 
I didn't go thorough all use cases, but just wondering if such possibility has been considered.

Reply:
I made some modifications so it is more clear which `this` is being used by `getMethodSet`.

Reply:
`infoCache` seems not for late use after current `getMethodHelper()` call hence won't help much performance.
In that case, `getInterfaces()` is a better choice.

Reply:
@theresa-m is there a PR update since last comment?

Reply:
In my opinion getMethodSet is okay. getInterfaces could be used but getMethodSet is already getting interfaces recursively calling getInterfaces. From a maintainability standpoint I think it will be advantageous to maintain one spot for the override logic in update: https://github.com/eclipse/openj9/blob/1ed4919d1a2f9eb58ce7bd4facc99bcb76068bf9/jcl/src/java.base/share/classes/java/lang/Class.java#L3890

Reply:
An alternative solution as discussed offline would be to move more of this logic into native code however this would require quite a bit of work.